llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1421 @@
+			if (getDataLayout().getPreferredAlign(GV) == 4)
+				return BitMask::ALIGNED4;
+			if (getDataLayout().getPreferredAlign(GV) == 2)
+				return BitMask::ALIGNED2;
+		}
+		if (const ConstantExpr* CE = dyn_cast<ConstantExpr>(V))
+		{
+			if (CE->getOpcode() == Instruction::GetElementPtr)
+			{
+				const GEPOperator *GEP = cast<GEPOperator>(CE);
+				if (GEP->hasAllZeroIndices())
+					return getBitMask(CE->getOperand(0));
+			}
+		}
+
+		auto it = stronglyKnownBits.back().find(V);
+		if (it != stronglyKnownBits.back().end())
+			return it->second;
+
+		return BitMask::NONE;
+	}
+	bool isValueComputed(const llvm::Value* V) const
+	{
+		if (isValueComputedConstant(V))
+			return true;
+		if (isa<BasicBlock>(V))
+			return true;
+		if (isa<Argument>(V) || isa<Instruction>(V))
+		{
+			return stronglyKnownBits.back().count(V);
+		}
+		return false;
+	}
+	bool areOperandsComputed(const llvm::Instruction& I)
+	{
+		if (isa<SelectInst>(I))
+		{
+			if (isValueComputed(I.getOperand(0)))
+			{
+				GenericValue V = getOperandValue(I.getOperand(0));
+				if (V.IntVal == 0u)
+				{
+					if (isValueComputed(I.getOperand(2)))
+						return true;
+				}
+				else
+				{
+					if (isValueComputed(I.getOperand(1)))
+						return true;
+				}
+
+			}	
+		}
+		for (auto& op : I.operands())
+		{
+			if (!isValueComputed(op))
+				return false;
+		}
+		if (const SwitchInst* SI = dyn_cast <SwitchInst>(&I))
+		{
+			if ((getBitMask(SI->getCondition()) != BitMask::ALL) )
+				return false;
+		}
+		if (const BranchInst* SI = dyn_cast <BranchInst>(&I))
+		{
+			if (SI->isConditional())
+				if ((getBitMask(SI->getCondition()) != BitMask::ALL) )
+					return false;
+		}
+		return true;
+	}
+	bool hasToBeSkipped(const llvm::Instruction& I)
+	{
+		if (!areOperandsComputed(I))
+			return true;
+
+		switch (I.getOpcode())
+		{
+			case Instruction::LandingPad:
+			case Instruction::Invoke:
+			case Instruction::ExtractElement:
+			case Instruction::CatchPad:
+			case Instruction::CleanupPad:
+			case Instruction::CleanupRet:
+			case Instruction::Fence:
+			case Instruction::Unreachable:
+			case Instruction::InsertValue:
+			case Instruction::InsertElement:
+			case Instruction::Store:
+			case Instruction::VAArg:
+			{
+				return true;
+			}
+			case Instruction::Call:
+			{
+				const CallBase& CB = cast<CallBase>(I);
+				if (const Function* FF = dyn_cast<Function>(CB.getCalledFunction()))
+				{
+					if (isAlreadyInCallStack(FF))
+						break;
+
+					if (FF->isDeclaration())
+						break;
+
+					std::vector<std::pair<const llvm::Value*, BitMask>> knownArgs;
+
+					uint32_t argIndex=0;
+					for (auto& op : CB.args())
+					{
+						if (argIndex >= FF->getFunctionType()->getNumParams())
+						{
+							//It means it's a VAArg, break from the inner loop
+							break;
+						}
+						if (isValueComputed(op))
+						if (FF->getArg(argIndex))
+							knownArgs.push_back({FF->getArg(argIndex), getBitMask(op)});
+
+						argIndex++;
+					}
+
+					addStackFrame();
```

Comment:
That 'hasToBeSkipped" also populate the stack frame does not make much sense. This method should only make decision, not act on them.

Reply:
ok, will split the decision vs populating values